rom/dev/src/flow/debug_unlock.rs

```diff
@@ -0,0 +1,315 @@
+        let challenge = env.trng.generate()?.as_bytes().try_into().unwrap();
+        let challenge_resp = ProductionAuthDebugUnlockChallenge {
+            vendor_id: request.vendor_id,
+            object_data_type: request.object_data_type,
+            length: length.to_ne_bytes()[..3].try_into().unwrap(),
+            unique_device_identifier: {
+                let mut id = [0u8; 32];
+                id[..17].copy_from_slice(&env.soc_ifc.fuse_bank().ueid());
+                id
+            },
+            challenge,
+            ..Default::default()
+        };
+        txn.send_response(challenge_resp.as_bytes())?;
+        core::mem::drop(txn);
+
+        // AUTH_DEBUG_UNLOCK_TOKEN
+        let txn = loop {
+            // Random delay for CFI glitch protection.
+            CfiCounter::delay();
+
+            match mbox.peek_recv() {
+                Some(txn) => break txn,
+                None => continue,
+            }
+        };
+
+        if CommandId::from(txn.cmd()) != CommandId::PRODUCTION_AUTH_DEBUG_UNLOCK_TOKEN {
+            Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_TOKEN_MBOX_CMD)?
+        }
+
+        env.soc_ifc.set_ss_dbg_unlock_in_progress(true);
+
+        let mut txn = txn.start_txn();
+        let mut request = ProductionAuthDebugUnlockToken::default();
+        FirmwareProcessor::copy_req_verify_chksum(&mut txn, request.as_bytes_mut())?;
+
+        let result: CaliptraResult<()> = (|| {
+            // Validate payload
+            if payload_length(request.length)
+                != size_of::<ProductionAuthDebugUnlockToken>() - size_of::<MailboxReqHeader>()
+            {
+                Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_TOKEN)?
+            }
+            // Debug level
+            if payload_length(request.unlock_category) != dbg_level {
+                Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_TOKEN)?
+            }
+            if request.challenge != challenge {
+                Err(CaliptraError::ROM_SS_DBG_UNLOCK_PROD_INVALID_TOKEN)?
+            }
+
+            let debug_auth_pk_offset = env
+                .soc_ifc
+                .debug_unlock_pk_hash_offset(payload_length(request.unlock_category))?
+                as u64;
+            let mci_base = env.soc_ifc.ss_mci_axi_base();
+            let debug_auth_pk_hash_base = mci_base + debug_auth_pk_offset.into();
+
+            let dma = &mut env.dma;
+            let mut fuse_digest = Array4x16::default();
+            dma.read_buffer(debug_auth_pk_hash_base, fuse_digest.as_bytes_mut())?;
+
+            let mut digest_op = env.sha2_512_384.sha512_digest_init()?;
+            digest_op.update(&request.ecc_public_key)?;
+            digest_op.update(&request.mldsa_public_key)?;
+            let mut request_digest = Array4x16::default();
+            digest_op.finalize(&mut request_digest)?;
+
+            // Verify that digest of keys match
+            if cfi_launder(request_digest) != fuse_digest {
```

Comment:
It may not be strictly necessary to use constant-time equality for this digest, but I'd like to anyway.